### PR TITLE
feat: rewrote parser.js to be iterative

### PR DESCRIPTION
**Pull request by simlu**

### CODE REVIEW

File: src/util/parser.js [hunk 1]
```diff
@@ -1,21 +1,142 @@
...
+const PARENT = Symbol("parent");
+const setParent = (input, parent) => Object.defineProperty(input, PARENT, { value: parent, writable: true });
+const getParent = input => (input[PARENT] === undefined ? null : input[PARENT]);
+
+const OR = Symbol("or");
+const markOr = input => Object.defineProperty(input, OR, { value: true, writable: false });
+const isOr = input => (input[OR] === true);
+module.exports.isOr = isOr;
+
+const Result = (input) => {
```
Comment (MacMcIrish) [inline src/util/parser.js]: What about `const Parser`? For clarity

File: src/util/parser.js [hunk 2]
```diff
@@ -1,21 +1,142 @@
...
+  const throwError = (msg, context = {}) => {
```
Comment (MacMcIrish) [inline src/util/parser.js]: Move outside of initialisation?

File: src/util/parser.js (hunk 1 above)
Comment (simlu) [reply src/util/parser.js]: This encapsulates the result. It doesn't actually do the parsing.

File: src/util/parser.js (hunk 2 above)
Comment (simlu) [reply src/util/parser.js]: Yeah, I was thinking about it already.

**Pull request merged**